File: mock-device-for-mqtt.c
/*
 * 1: 依赖库：cjson，eclipse paho mqtt.编译前请安装
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <pthread.h>
#include <signal.h>
#include <cjson/cJSON.h>
#include "MQTTClient.h"

#define ADDRESS     "tcp://localhost:1883"
#define CLIENTID    "Mock-Device-Client"
#define DATA_TOPIC  "DataTopic"
#define CMDTOPIC    "CommandTopic"
#define RESPONSE_TOPIC "ResponseTopic"
#define USERNAME    "huaqiao"
#define PWD         "1234"
#define PAYLOAD     "{\"name\":\"mqtt-device-01\",\"randnum\":\"520.1314\"}"
#define QOS         0
#define TIMEOUT     10000L

static pthread_mutex_t mutex;

char *active = "false";

void publish(MQTTClient *client, char *payload, char *topic) {
  MQTTClient_message pubmsg = MQTTClient_message_initializer;
  MQTTClient_deliveryToken token;

  pubmsg.payload = payload;
  pubmsg.payloadlen = strlen(payload);
  pubmsg.qos = QOS;
  pubmsg.retained = 0;
  MQTTClient_publishMessage(*client, topic, &pubmsg, &token);
  printf("Waiting for up to %d seconds for publication of %s\n"
          "on topic %s for client with ClientID: %s\n",
          (int)(TIMEOUT/1000), payload, topic, CLIENTID);
  MQTTClient_waitForCompletion(*client, token, TIMEOUT);
  printf("Message with delivery token %d delivered\n", token);
}

void subscribe(MQTTClient *client) {
  MQTTClient_subscribe(*client, CMDTOPIC, QOS);
  printf("Subscribing to topic %s\nfor client %s using QoS%d\n\n", CMDTOPIC, CLIENTID, QOS);
}

void response_cmd(MQTTClient *client, char *resp_data) {
    publish(client,resp_data,RESPONSE_TOPIC);
}

int on_message(void *context, char *topicName, int topicLen, MQTTClient_message *message) {
    char *cmd;
    char *method;
    char *param;
    cJSON *item = NULL;
    cJSON *json = NULL;
    printf("Message arrived\n");
    printf("     topic: %s\n", topicName);
    printf("   message: %s\n",message->payload);

    json = cJSON_Parse(message->payload);
    item = cJSON_GetObjectItem(json,"cmd");

    cmd = item->valuestring;
    method = (cJSON_GetObjectItem(json,"method"))->valuestring;
    //处理ping命令
    if (strcmp(cmd,"ping") == 0) {
      cJSON_AddItemToObject(json, "ping", cJSON_CreateString("pong"));
    }

    //处理message命令
    if (strcmp(cmd,"message") == 0) {
      if (strcmp(method,"get") == 0) {
        cJSON_AddItemToObject(json, "message", cJSON_CreateString("Are you ok?"));
      } else {
       //param = cJSON_Print(cJSON_GetObjectItem(json,"param"));
       cJSON_AddItemToObject(json, "result", cJSON_CreateString("set success."));
      }
    }
    //处理randnum命令
    if (strcmp(cmd,"randnum") == 0) {
      cJSON_AddItemToObject(json, "randnum", cJSON_CreateString("520.1314"));
    }
    //处理collect命令
    if (strcmp(cmd,"collect") == 0) {
      if (strcmp(method,"get") == 0) {
        cJSON_AddItemToObject(json, "collect", cJSON_CreateString(active));
      } else {
        active = (cJSON_GetObjectItem(json,"param"))->valuestring;
      }
    }

    response_cmd((MQTTClient *)context,cJSON_PrintUnformatted(json));
    MQTTClient_freeMessage(&message);
    MQTTClient_free(topicName);
    // cJSON_Delete(json);
    // cJSON_Delete(item);
    return 1;
}

void *send_data_actively_server (void *context) {
  for (;;) {
    //pthread_mutex_lock(&mutex);
    if (strcmp(active,"true") == 0) {
      sleep(1);
      printf("send data actively from mock device.\n");
      publish((MQTTClient *)context,PAYLOAD,DATA_TOPIC);
    }
    //pthread_mutex_unlock(&mutex);
  }
}

int main(int argc, char* argv[]) {
    MQTTClient client;
    MQTTClient_create(&client, ADDRESS, CLIENTID,  MQTTCLIENT_PERSISTENCE_NONE, NULL);
    MQTTClient_connectOptions conn_opts = MQTTClient_connectOptions_initializer;
    conn_opts.username = USERNAME;
    conn_opts.password = PWD;
    conn_opts.keepAliveInterval = 20;
    conn_opts.cleansession = 1;
    int rc;
    pthread_t thread_id;
    MQTTClient_setCallbacks(client, &client, NULL, on_message, NULL);
    if ((rc = MQTTClient_connect(client, &conn_opts)) != MQTTCLIENT_SUCCESS) {
        printf("Failed to connect, return code %d\n", rc);
        exit(EXIT_FAILURE);
    } else {
      printf("Successful connection\n");
    }
    subscribe(&client);
    pthread_create(&thread_id, NULL, send_data_actively_server, (void *)&client);
    // pthread_join(thread_id, NULL);
    for (;;) {
      sleep(3);
    }
    MQTTClient_disconnect(client, 10000);
    MQTTClient_destroy(&client);
    return rc;
}

File: jakarta-v2.1.0/mock-device-for-mqtt.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

#version: python 2.7
#如果是python3 需要修改部分依赖包或者语句语法
#用于模拟一个基于mqtt协议传输虚拟设备驱动

import paho.mqtt.client as mqtt
import json
import time
import Queue
import threading,random

BROKER_HOST_ADDR   = "localhost"
BROKER_HOST_PORT   = 1883
USERNAME    = "huaqiaoz" #非安全模式忽略用户名秘密
PWD         = "1234" #非安全模式忽略用户名秘密
#cmd topic本质上就是你的设备监听的topic，
#也是在UI上添加device的时候，地址中所填数据，和用户名密码等一起组成当前设备的唯一标识。
CMD_TOPIC   = "CommandTopic"
RESPONSE_TOPIC = "ResponseTopic"
DATA_TOPIC  = "DataTopic"

globalQueue = Queue.Queue()

def gen():
       return round(random.uniform(0, 50),2)

def send_data():
    #java版本, name的值为添加的设备名
    #data = {"randnum":520.1314,"name":"mqtt-device-01"}

    #go版本, name的值为添加的设备名（required）, go版本的区别是必须带上cmd字段
    # data = {"randnum":520.1314,"name":"mqtt-device-01","cmd":"randnum","method":"get"}
    
    #v2.1.0版本数据格式
    #数据格式 data = {"randnum":520.1314,"name":"mqtt-device-01","cmd":"randnum","method":"get"}
    data = {"randnum":gen(),"name":"mqtt-device-01","cmd":"randnum", "method":"get"}
    print("sending data actively! " + json.dumps(data))
    client.publish(DATA_TOPIC,json.dumps(data) , qos=0, retain=False)

class SendDataActiveServer(threading.Thread):
    def __init__(self,threadID,name,queue):
        super(SendDataActiveServer,self).__init__()
        self.threadID = threadID
        self.name = name
        self.queue = queue
        self.active = False

    def run(self):
        while 1==1 :
          #v2.1.0 接收到的bool值自动自动转换为python能识别的bool值，无须二次转换
          if self.active:
             send_data()
             time.sleep(1)
             self.getItemFromQueue()
          else:
             time.sleep(1)
             self.getItemFromQueue()

    def getItemFromQueue(self):
        try:
          if self.queue.get(block=False):
             self.active = True
          else:
             self.active = False
        except Queue.Empty:
          #quene.get()方法在队列中为空是返回异常，捕获异常什么都不做，保持active原状
          time.sleep(0.1)

#当接收到命令，响应命令
#v2.1.0 get命令格式： {"cmd":"ping","method":"get","uuid":"f46ae5c7-2a08-4f56-a38b-d9e122d255d0"}
#v2.1.0 set命令格式 {"cmd":"ping","ping":"ping_set_value","method":"set","uuid":"95d559ad-7140-4d62-8dd8-6465c37fec10"}
def on_message(client, userdata, msg):
    print(msg.topic+" "+str(msg.payload)+'\n')
    d = json.loads(msg.payload)

    if d['cmd'] == "message":
       print("This is message cmd")
       if d['method'] == "get":
          d['message'] = "Are you ok?"
       elif d['method'] == "set":
          d['result'] = "set successed."

    if d['cmd'] == "ping":
       print("This is ping cmd")
       d['ping'] = "pong"

    if d['cmd'] == "randnum":
       print("This is randnum cmd")
       d['randnum'] = gen()

    if d['cmd'] == "collect" and d['method'] == "set":
       print("This is collect set cmd")
       d['result'] = "set successed."
       #param的值是true或false,且是字符串类型, 
       #globalQueue.put(d['param'])
       #v2.1.0命令格式中移除了param属性，使用device resource name为实际参数名
       globalQueue.put(d['collect'])
    elif d['cmd'] == "collect" and d['method'] == "get":
       print("This is collect get cmd")
       d['collect'] = thread.active

    print(json.dumps(d))
    client.publish(RESPONSE_TOPIC, json.dumps(d))

def on_connect(client, userdata, flags, rc):
    print("Connected success with result code "+str(rc))
    #监听命令
    client.subscribe(CMD_TOPIC)

client = mqtt.Client()
client.username_pw_set(USERNAME, PWD)
client.on_message = on_message
client.on_connect = on_connect

client.connect(BROKER_HOST_ADDR, BROKER_HOST_PORT, 60)

#开始独立线程用于主动发送数据
thread = SendDataActiveServer("Thread-1", "SendDataServerThread", globalQueue)
thread.setDaemon(True)
thread.start()

client.loop_forever()
